### Business Analytics - Assignment 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 4 (18 August) @ 11.59pm  
**Submission**: Submit the assignment using the submission link on iLearn


- Put **all your work** into this file;
- Solutions must be provided in the cells below each question. Do not insert new cells; 
- Failure to follow these instructions will result in a loss of 20 points.


### About the Assignment

Credit score cards are used as a risk control method in the financial industry. Personal information submitted by credit card applicants are used to predict the probability of future defaults. The bank employs such data to decide whether to issue a credit card to the applicant or not.




| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| Income   | Annual income  |  |
| Gender   | Applicant's Gender   | Male = 0, Female = 1  |
| Car | Car Ownership | Yes = 1, No = 0 | 
| Children | Number of Children | |
| Real Estate | Real Estate Ownership | Yes = 1, No = 0 
| Days Since Birth | No. of Days | Count backwards from current day (0), -1 means yesterday
| Days Employed | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| Payment Default | Whether a client has overdue credit card payments | Yes = 1, No = 0



In [28]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Perceptron
from sklearn.linear_model import LogisticRegression


### Problem 1 - (50 points) {-}


**Question 1** 

- Import the `assignment_data.xlsx` file from `data` folder into a pandas DataFrame named `df`; 
- Delete duplicate rows from `df` according to `ID`;
- Delete the `ID` column.
- How many rows are left in `df`? (answer in the Markdown box below in a full sentence)

(10 points)

In [1]:
# --------- add your code in this cell ---------
df = pd.read_excel('C:/Users/PC/Desktop/KietAUS-Assignment1/data/assignment_data.xlsx')
df

df = df.drop_duplicates(subset='ID', keep='first')
df

df = df.drop(columns=['ID'])
df


Gender  Car  Real Estate  Children    Income  Days Since Birth  \
0          1    1            1         1  135000.0            -10434   
1          1    0            0         0  157500.0            -22494   
2          1    0            1         0  675000.0            -17964   
3          1    0            0         0  193500.0            -17880   
4          0    1            0         0  270000.0            -20776   
...      ...  ...          ...       ...       ...               ...   
9993       1    1            0         0   81000.0            -17493   
9994       1    0            1         0   45000.0            -22072   
9995       1    0            0         0  306000.0            -12677   
9996       1    0            0         0  121500.0            -23491   
9998       1    0            1         0  144000.0            -19002   

      Days Employed  Payment Default  
0              -803                0  
1            365243                1  
2             -6594                0  
3             -9858                1  
4             -2745                1  
...             ...              ...  
9993          -2347                0  
9994         365243                0  
9995          -5587                1  
9996         365243                0  
9998         365243                0  

[6805 rows x 8 columns]

------- add your text answer in this cell ----------
There are 6805 rows left in df

---

**Question 2**

- Reset the index in `df` using an appropriate function from `pandas` so that the new index corresponds to the number of rows (make sure to delete the old index). 
- How many positive values of `Days Employed` are there? (answer in Markdown)
- Replace the positive values of `Days Employed` with 0 (zero) in `df`

(10 points)

In [9]:
# --------- add your code in this cell ---------
df = df.reset_index(drop=True)

print ("Positive values of Days Employed are:", df[df['Days Employed'] > 0].shape[0])

column_name = 'Days Employed'
df[column_name] = df['Days Employed'].apply(lambda x: 0 if x > 0 else x)
df

Positive values of Days Employed are: 1069


Gender  Car  Real Estate  Children    Income  Days Since Birth  \
0          1    1            1         1  135000.0            -10434   
1          1    0            0         0  157500.0            -22494   
2          1    0            1         0  675000.0            -17964   
3          1    0            0         0  193500.0            -17880   
4          0    1            0         0  270000.0            -20776   
...      ...  ...          ...       ...       ...               ...   
6800       1    1            0         0   81000.0            -17493   
6801       1    0            1         0   45000.0            -22072   
6802       1    0            0         0  306000.0            -12677   
6803       1    0            0         0  121500.0            -23491   
6804       1    0            1         0  144000.0            -19002   

      Days Employed  Payment Default  
0              -803                0  
1                 0                1  
2             -6594                0  
3             -9858                1  
4             -2745                1  
...             ...              ...  
6800          -2347                0  
6801              0                0  
6802          -5587                1  
6803              0                0  
6804              0                0  

[6805 rows x 8 columns]

------- add your text answer in this cell ----------
There are 1069 Positive values of Days Employed.


---
**Question 3**

Create two new variables in `df` named 

1. `Age`;
2. `Years in Employment`,

which measure age and employment length in **years** (positive decimal numbers) from `Days Since Birth` and `Days Employed` by applying appropriate transformations on these variables. 

Delete the original variables `Days Since Birth` and `Days Employed`.

(5 points)


In [11]:
# --------- add your code in this cell ---------
df['Age'] = df['Days Since Birth'] / 365
df['Years in Employment'] = df['Days Employed'] / 365
df

df.drop(['Days Since Birth', 'Days Employed'], axis=1, inplace=True)
df

Gender  Car  Real Estate  Children    Income  Payment Default  \
0          1    1            1         1  135000.0                0   
1          1    0            0         0  157500.0                1   
2          1    0            1         0  675000.0                0   
3          1    0            0         0  193500.0                1   
4          0    1            0         0  270000.0                1   
...      ...  ...          ...       ...       ...              ...   
6800       1    1            0         0   81000.0                0   
6801       1    0            1         0   45000.0                0   
6802       1    0            0         0  306000.0                1   
6803       1    0            0         0  121500.0                0   
6804       1    0            1         0  144000.0                0   

            Age  Years in Employment  
0    -28.586301            -2.200000  
1    -61.627397             0.000000  
2    -49.216438           -18.065753  
3    -48.986301           -27.008219  
4    -56.920548            -7.520548  
...         ...                  ...  
6800 -47.926027            -6.430137  
6801 -60.471233             0.000000  
6802 -34.731507           -15.306849  
6803 -64.358904             0.000000  
6804 -52.060274             0.000000  

[6805 rows x 8 columns]

---
**Question 4**

- Create a **one**-dimensional NumPy array named `y` by exporting the first 5,000 observations of `Payment_Default`. (Hint: see `ravel()` function)
- Create a NumPy array named `X` by exporting the first 5,000 observations of the following columns `Gender`, `Car`, `Real Estate`, `Children`, `Income`, `Age`, `Years in Employment`.
 
(10 points)


In [18]:
# --------- add your code in this cell ---------
y = df['Payment Default'].iloc[:4999].values.ravel()
selected_columns = ['Gender', 'Car', 'Real Estate', 'Children', 'Income', 'Age', 'Years in Employment']
X = df[selected_columns].iloc[:4999].values

print (y)
print (X)

[0 1 0 ... 1 1 1]
[[ 1.00000000e+00  1.00000000e+00  1.00000000e+00 ...  1.35000000e+05
  -2.85863014e+01 -2.20000000e+00]
 [ 1.00000000e+00  0.00000000e+00  0.00000000e+00 ...  1.57500000e+05
  -6.16273973e+01  0.00000000e+00]
 [ 1.00000000e+00  0.00000000e+00  1.00000000e+00 ...  6.75000000e+05
  -4.92164384e+01 -1.80657534e+01]
 ...
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  2.02500000e+05
  -2.88356164e+01 -2.36438356e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  9.90000000e+04
  -4.89753425e+01 -1.43534247e+01]
 [ 1.00000000e+00  0.00000000e+00  1.00000000e+00 ...  1.57500000e+05
  -5.01726027e+01 -1.35506849e+01]]


---

**Question 5** 

- Use an appropriate `scikit-learn` library we learned in class to create the following NumPy arrays: `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% training and 30% test datasets. 
- Set `random_state` to 0 and stratify subsamples so that train and test datasets have roughly equal proportions of the target's class labels. 

(5 points) 

In [19]:
# --------- add your code in this cell ---------
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0, stratify = y)


---

**Question 6**

- Create new variables by using an appropriate `scikit-learn` library we learned in class to standardize the features from the training and test datasets to mean zero and variance one. Name the new variables by appending '_scaled' to the original variable names.


(10 points)   

In [24]:
# --------- add your code in this cell ---------
# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler to the training data and transform the training data
X_train_scaled = scaler.fit(X_train).transform(X_train)

# Transform the test data using the same scaler
X_test_scaled = scaler.transform(X_test)

# Create a list of new column names for the scaled features
selected_columns_scaled = [col + '_scaled' for col in selected_columns]

# Create dataframes for scaled training and test data
X_train_scaled_df = pd.DataFrame(
    X_train_scaled, columns=selected_columns_scaled)
X_test_scaled_df = pd.DataFrame(X_test_scaled, columns=selected_columns_scaled)

# Print the scaled dataframes
print(X_train_scaled_df)
print(X_test_scaled_df)

      Gender_scaled  Car_scaled  Real Estate_scaled  Children_scaled  \
0         -1.408172    1.267028            0.731657        -0.519164   
1          0.710140    1.267028            0.731657         1.783188   
2          0.710140    1.267028            0.731657        -0.519164   
3          0.710140   -0.789248            0.731657        -0.519164   
4          0.710140   -0.789248            0.731657        -0.519164   
...             ...         ...                 ...              ...   
3494      -1.408172   -0.789248            0.731657        -0.519164   
3495       0.710140   -0.789248           -1.366760         0.632012   
3496      -1.408172    1.267028            0.731657        -0.519164   
3497       0.710140   -0.789248           -1.366760         0.632012   
3498       0.710140   -0.789248            0.731657         1.783188   

      Income_scaled  Age_scaled  Years in Employment_scaled  
0          0.099801   -1.604409                   -0.705783  
1          

---

## Problem 2 - (20 Points) {-}

**Question 7**

Fit the following two classifiers to the transformed training dataset using `scikit-learn` libraries.

- Perceptron - name your instance `pc` set `random_state=7`
- Logistic Regression - name your instance `lr` set `random_state=7`

When initializing instances of the above classifiers only set the parameters referenced above and nothing else.

(20 points)

In [30]:
# Initialize the Perceptron classifier
pc = Perceptron(random_state=7)

# Fit the Perceptron on the training data
pc.fit(X_train_scaled, y_train)


Perceptron(random_state=7)

In [31]:
# Initialize the Logistic Regression classifier
lr = LogisticRegression(random_state=7)

# Fit the Logistic Regression on the training data
lr.fit(X_train_scaled, y_train)

LogisticRegression(random_state=7)

---
## Problem 3 - (30 points) {-}


**Question 8**

- Using a method built into each of the two classifiers compute their prediction accuracies on the training data;
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_train`, e.g. you should have `lr_accuracy_train`; 
- Print the two accuracy **variables** along with their brief descriptions.

(5 points)

In [32]:
pc_accuracy_train = pc.score(X_train_scaled, y_train)
lr_accuracy_train = lr.score(X_train_scaled, y_train)

# Print the accuracy values along with their descriptions
print("Perceptron Accuracy on Training Data:", pc_accuracy_train)
print("Logistic Regression Accuracy on Training Data:", lr_accuracy_train)

Perceptron Accuracy on Training Data: 0.5324378393826807
Logistic Regression Accuracy on Training Data: 0.6093169476993426


---

**Question 9** 

- Using a method built into each of the above classifiers compute their prediction accuracy for the test dataset
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_test`, e.g. you should have `lr_accuracy_test`. 
- Print the two accuracy **variables** along with brief descriptions.

(5 points)

In [33]:
# Compute the prediction accuracies on the test data
pc_accuracy_test = pc.score(X_test_scaled, y_test)
lr_accuracy_test = lr.score(X_test_scaled, y_test)

# Print the accuracy values along with their descriptions
print("Perceptron Accuracy on Test Data:", pc_accuracy_test)
print("Logistic Regression Accuracy on Test Data:", lr_accuracy_test)

Perceptron Accuracy on Test Data: 0.5453333333333333
Logistic Regression Accuracy on Test Data: 0.62


---

**Question 10** 

Using nicely formated text in Markdown comment on the accuracies computed in Questions 8 & 9 making sure you address:
1. Training vs test set datasets for each classifier; (5 marks)
2. Perceptrion vs Logistic Regression. (5 marks) 
3. Are the results as expected, and why or why not? (Hint: You are not expected to comment on why a particular model is better. 10 marks) 

(20 marks)


## Question 10: Commentary on Classifier Accuracies

### Training vs Test Set Performance

When analyzing the accuracy values computed for the Perceptron and Logistic Regression classifiers, it's important to consider their performance on both the training and test datasets.

For the Perceptron classifier, the accuracy on the training data is approximately 53.24%, while on the test data, it is around 54.53%. This indicates that the Perceptron is performing slightly better on the test set compared to the training set. Similarly, for the Logistic Regression classifier, the accuracy on the training data is approximately 60.93%, and on the test data, it is about 62.00%. Like the Perceptron, the Logistic Regression classifier also shows a slightly better performance on the test set.

### Perceptron vs Logistic Regression

Comparing the accuracy values between the Perceptron and Logistic Regression classifiers, it's evident that the Logistic Regression classifier is consistently outperforming the Perceptron in terms of accuracy on both the training and test datasets.

On the training data, the Logistic Regression accuracy is notably higher (60.93%) compared to the Perceptron accuracy (53.24%). Similarly, on the test data, the Logistic Regression accuracy (62.00%) surpasses the Perceptron accuracy (54.53%).

### Expected Results

The observed results align with expectations to some extent. Logistic Regression is a more complex and flexible model compared to the Perceptron. This allows Logistic Regression to capture more intricate relationships within the data, leading to higher accuracy. However, it's important to note that accuracy alone might not provide a complete picture of model performance. The slight improvement of accuracy on the test set compared to the training set suggests that both classifiers are not significantly overfitting.

Overall, the differences in accuracy between the two classifiers are in line with their inherent characteristics. Logistic Regression's ability to handle non-linear relationships and its overall flexibility make it a stronger candidate for this classification task compared to the simpler Perceptron model.

Please note that while the Logistic Regression model outperforms the Perceptron in terms of accuracy, this commentary does not delve into the reasons why Logistic Regression might be inherently better for this specific task. Such an analysis would require a deeper understanding of the dataset, features, and problem context.


---
---

### Marking Criteria

- To receive full marks your solutions must satisfy the following criteria:
    - Provide numerically correct answers;
    - Use the specific Python methods and libraries asked for in each question;
    - Written answers (where required) must be correct and formulated in easy to understand full sentences. 

---
---